File: scala-akka-http-server/src/main/scala/org/openapitools/server/api/StreamsApi.scala
package org.openapitools.server.api

import akka.http.scaladsl.server.Directives._
import akka.http.scaladsl.server.Route
import akka.http.scaladsl.model.StatusCodes
import akka.http.scaladsl.marshalling.ToEntityMarshaller
import akka.http.scaladsl.unmarshalling.FromEntityUnmarshaller
import akka.http.scaladsl.unmarshalling.FromStringUnmarshaller
import org.openapitools.server.AkkaHttpHelper._
import org.openapitools.server.model.CodedError
import org.openapitools.server.model.JsonError
import org.openapitools.server.model.JsonSuccess
import org.openapitools.server.model.JsonSuccessBase
import org.openapitools.server.model.NonExistingStreamError
import org.openapitools.server.model.OneOfobjectobject
import org.openapitools.server.model.OneOfstringinteger


class StreamsApi(
    streamsService: StreamsApiService,
    streamsMarshaller: StreamsApiMarshaller
) {

  
  import streamsMarshaller._

  lazy val route: Route =
    path("streams" / IntNumber) { (streamId) => 
      delete {  
            streamsService.archiveStream(streamId = streamId)
      }
    } ~
    path("calls" / "bigbluebutton" / "create") { 
      get {  
            streamsService.createBigBlueButtonVideoCall()
      }
    } ~
    path("streams" / IntNumber / "delete_topic") { (streamId) => 
      post { 
        parameters("topic_name".as[String]) { (topicName) => 
            streamsService.deleteTopic(streamId = streamId, topicName = topicName)
        }
      }
    } ~
    path("get_stream_id") { 
      get { 
        parameters("stream".as[String]) { (stream) => 
            streamsService.getStreamId(stream = stream)
        }
      }
    } ~
    path("users" / "me" / IntNumber / "topics") { (streamId) => 
      get {  
            streamsService.getStreamTopics(streamId = streamId)
      }
    } ~
    path("streams") { 
      get { 
        parameters("include_public".as[Boolean].?(true), "include_web_public".as[Boolean].?(false), "include_subscribed".as[Boolean].?(true), "include_all_active".as[Boolean].?(false), "include_default".as[Boolean].?(false), "include_owner_subscribed".as[Boolean].?(false)) { (includePublic, includeWebPublic, includeSubscribed, includeAllActive, includeDefault, includeOwnerSubscribed) => 
            streamsService.getStreams(includePublic = includePublic, includeWebPublic = includeWebPublic, includeSubscribed = includeSubscribed, includeAllActive = includeAllActive, includeDefault = includeDefault, includeOwnerSubscribed = includeOwnerSubscribed)
        }
      }
    } ~
    path("streams" / IntNumber / "members") { (streamId) => 
      get {  
            streamsService.getSubscribers(streamId = streamId)
      }
    } ~
    path("users" / IntNumber / "subscriptions" / IntNumber) { (userId, streamId) => 
      get {  
            streamsService.getSubscriptionStatus(userId = userId, streamId = streamId)
      }
    } ~
    path("users" / "me" / "subscriptions") { 
      get { 
        parameters("include_subscribers".as[Boolean].?(false)) { (includeSubscribers) => 
            streamsService.getSubscriptions(includeSubscribers = includeSubscribers)
        }
      }
    } ~
    path("users" / "me" / "subscriptions" / "muted_topics") { 
      patch { 
        parameters("stream".as[String].?, "stream_id".as[Int].?, "topic".as[String], "op".as[String]) { (stream, streamId, topic, op) => 
            streamsService.muteTopic(topic = topic, op = op, stream = stream, streamId = streamId)
        }
      }
    } ~
    path("users" / "me" / "subscriptions") { 
      post { 
        parameters("subscriptions".as[String], "principals".as[String].?("new ListBuffer[OneOfstringinteger]() "), "authorization_errors_fatal".as[Boolean].?(true), "announce".as[Boolean].?(false), "invite_only".as[Boolean].?(false), "history_public_to_subscribers".as[Boolean].?, "stream_post_policy".as[Int].?(1), "message_retention_days".as[String].?) { (subscriptions, principals, authorizationErrorsFatal, announce, inviteOnly, historyPublicToSubscribers, streamPostPolicy, messageRetentionDays) => 
            streamsService.subscribe(subscriptions = subscriptions, principals = principals, authorizationErrorsFatal = authorizationErrorsFatal, announce = announce, inviteOnly = inviteOnly, historyPublicToSubscribers = historyPublicToSubscribers, streamPostPolicy = streamPostPolicy, messageRetentionDays = messageRetentionDays)
        }
      }
    } ~
    path("users" / "me" / "subscriptions") { 
      delete { 
        parameters("subscriptions".as[String], "principals".as[String].?("new ListBuffer[OneOfstringinteger]() ")) { (subscriptions, principals) => 
            streamsService.unsubscribe(subscriptions = subscriptions, principals = principals)
        }
      }
    } ~
    path("streams" / IntNumber) { (streamId) => 
      patch { 
        parameters("description".as[String].?, "new_name".as[String].?, "is_private".as[Boolean].?, "is_announcement_only".as[Boolean].?, "stream_post_policy".as[Int].?(1), "history_public_to_subscribers".as[Boolean].?, "message_retention_days".as[String].?) { (description, newName, isPrivate, isAnnouncementOnly, streamPostPolicy, historyPublicToSubscribers, messageRetentionDays) => 
            streamsService.updateStream(streamId = streamId, description = description, newName = newName, isPrivate = isPrivate, isAnnouncementOnly = isAnnouncementOnly, streamPostPolicy = streamPostPolicy, historyPublicToSubscribers = historyPublicToSubscribers, messageRetentionDays = messageRetentionDays)
        }
      }
    } ~
    path("users" / "me" / "subscriptions" / "properties") { 
      post { 
        parameters("subscription_data".as[String]) { (subscriptionData) => 
            streamsService.updateSubscriptionSettings(subscriptionData = subscriptionData)
        }
      }
    } ~
    path("users" / "me" / "subscriptions") { 
      patch { 
        parameters("delete".as[String].?("new ListBuffer[String]() "), "add".as[String].?("new ListBuffer[Any]() ")) { (delete, add) => 
            streamsService.updateSubscriptions(delete = delete, add = add)
        }
      }
    }
}


trait StreamsApiService {

  def archiveStream200(responseJsonSuccess: JsonSuccess)(implicit toEntityMarshallerJsonSuccess: ToEntityMarshaller[JsonSuccess]): Route =
    complete((200, responseJsonSuccess))
  def archiveStream400(responseJsonError: JsonError)(implicit toEntityMarshallerJsonError: ToEntityMarshaller[JsonError]): Route =
    complete((400, responseJsonError))
  /**
   * Code: 200, Message: Success., DataType: JsonSuccess
   * Code: 400, Message: Bad request., DataType: JsonError
   */
  def archiveStream(streamId: Int)
      (implicit toEntityMarshallerJsonSuccess: ToEntityMarshaller[JsonSuccess], toEntityMarshallerJsonError: ToEntityMarshaller[JsonError]): Route

  def createBigBlueButtonVideoCall200(responseJsonSuccessBase: JsonSuccessBase)(implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route =
    complete((200, responseJsonSuccessBase))
  /**
   * Code: 200, Message: Success., DataType: JsonSuccessBase
   */
  def createBigBlueButtonVideoCall()
      (implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route

  def deleteTopic200(responseJsonSuccess: JsonSuccess)(implicit toEntityMarshallerJsonSuccess: ToEntityMarshaller[JsonSuccess]): Route =
    complete((200, responseJsonSuccess))
  def deleteTopic400(responseJsonError: JsonError)(implicit toEntityMarshallerJsonError: ToEntityMarshaller[JsonError]): Route =
    complete((400, responseJsonError))
  /**
   * Code: 200, Message: Success., DataType: JsonSuccess
   * Code: 400, Message: Error., DataType: JsonError
   */
  def deleteTopic(streamId: Int, topicName: String)
      (implicit toEntityMarshallerJsonSuccess: ToEntityMarshaller[JsonSuccess], toEntityMarshallerJsonError: ToEntityMarshaller[JsonError]): Route

  def getStreamId200(responseJsonSuccessBase: JsonSuccessBase)(implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route =
    complete((200, responseJsonSuccessBase))
  def getStreamId400(responseCodedError: CodedError)(implicit toEntityMarshallerCodedError: ToEntityMarshaller[CodedError]): Route =
    complete((400, responseCodedError))
  /**
   * Code: 200, Message: Success., DataType: JsonSuccessBase
   * Code: 400, Message: Bad request., DataType: CodedError
   */
  def getStreamId(stream: String)
      (implicit toEntityMarshallerCodedError: ToEntityMarshaller[CodedError], toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route

  def getStreamTopics200(responseJsonSuccessBase: JsonSuccessBase)(implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route =
    complete((200, responseJsonSuccessBase))
  def getStreamTopics400(responseJsonError: JsonError)(implicit toEntityMarshallerJsonError: ToEntityMarshaller[JsonError]): Route =
    complete((400, responseJsonError))
  /**
   * Code: 200, Message: Success., DataType: JsonSuccessBase
   * Code: 400, Message: Bad request., DataType: JsonError
   */
  def getStreamTopics(streamId: Int)
      (implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase], toEntityMarshallerJsonError: ToEntityMarshaller[JsonError]): Route

  def getStreams200(responseJsonSuccessBase: JsonSuccessBase)(implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route =
    complete((200, responseJsonSuccessBase))
  def getStreams400(responseCodedError: CodedError)(implicit toEntityMarshallerCodedError: ToEntityMarshaller[CodedError]): Route =
    complete((400, responseCodedError))
  /**
   * Code: 200, Message: Success., DataType: JsonSuccessBase
   * Code: 400, Message: Bad request., DataType: CodedError
   */
  def getStreams(includePublic: Boolean, includeWebPublic: Boolean, includeSubscribed: Boolean, includeAllActive: Boolean, includeDefault: Boolean, includeOwnerSubscribed: Boolean)
      (implicit toEntityMarshallerCodedError: ToEntityMarshaller[CodedError], toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route

  def getSubscribers200(responseJsonSuccessBase: JsonSuccessBase)(implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route =
    complete((200, responseJsonSuccessBase))
  def getSubscribers400(responseJsonError: JsonError)(implicit toEntityMarshallerJsonError: ToEntityMarshaller[JsonError]): Route =
    complete((400, responseJsonError))
  /**
   * Code: 200, Message: Success., DataType: JsonSuccessBase
   * Code: 400, Message: Bad request., DataType: JsonError
   */
  def getSubscribers(streamId: Int)
      (implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase], toEntityMarshallerJsonError: ToEntityMarshaller[JsonError]): Route

  def getSubscriptionStatus200(responseJsonSuccessBase: JsonSuccessBase)(implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route =
    complete((200, responseJsonSuccessBase))
  /**
   * Code: 200, Message: Success, DataType: JsonSuccessBase
   */
  def getSubscriptionStatus(userId: Int, streamId: Int)
      (implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route

  def getSubscriptions200(responseJsonSuccessBase: JsonSuccessBase)(implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route =
    complete((200, responseJsonSuccessBase))
  /**
   * Code: 200, Message: Success., DataType: JsonSuccessBase
   */
  def getSubscriptions(includeSubscribers: Boolean)
      (implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route

  def muteTopic200(responseJsonSuccess: JsonSuccess)(implicit toEntityMarshallerJsonSuccess: ToEntityMarshaller[JsonSuccess]): Route =
    complete((200, responseJsonSuccess))
  def muteTopic400(responseOneOfobjectobject: OneOfobjectobject)(implicit toEntityMarshallerOneOfobjectobject: ToEntityMarshaller[OneOfobjectobject]): Route =
    complete((400, responseOneOfobjectobject))
  /**
   * Code: 200, Message: Success., DataType: JsonSuccess
   * Code: 400, Message: Bad request., DataType: OneOfobjectobject
   */
  def muteTopic(topic: String, op: String, stream: Option[String], streamId: Option[Int])
      (implicit toEntityMarshallerOneOfobjectobject: ToEntityMarshaller[OneOfobjectobject], toEntityMarshallerJsonSuccess: ToEntityMarshaller[JsonSuccess]): Route

  def subscribe200(responseOneOfobjectobject: OneOfobjectobject)(implicit toEntityMarshallerOneOfobjectobject: ToEntityMarshaller[OneOfobjectobject]): Route =
    complete((200, responseOneOfobjectobject))
  def subscribe400(responseOneOfobjectobject: OneOfobjectobject)(implicit toEntityMarshallerOneOfobjectobject: ToEntityMarshaller[OneOfobjectobject]): Route =
    complete((400, responseOneOfobjectobject))
  /**
   * Code: 200, Message: Success., DataType: OneOfobjectobject
   * Code: 400, Message: Success., DataType: OneOfobjectobject
   */
  def subscribe(subscriptions: String, principals: String, authorizationErrorsFatal: Boolean, announce: Boolean, inviteOnly: Boolean, historyPublicToSubscribers: Option[Boolean], streamPostPolicy: Int, messageRetentionDays: Option[String])
      (implicit toEntityMarshallerOneOfobjectobject: ToEntityMarshaller[OneOfobjectobject]): Route

  def unsubscribe200(responseJsonSuccessBase: JsonSuccessBase)(implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route =
    complete((200, responseJsonSuccessBase))
  def unsubscribe400(responseNonExistingStreamError: NonExistingStreamError)(implicit toEntityMarshallerNonExistingStreamError: ToEntityMarshaller[NonExistingStreamError]): Route =
    complete((400, responseNonExistingStreamError))
  /**
   * Code: 200, Message: Success., DataType: JsonSuccessBase
   * Code: 400, Message: Bad request., DataType: NonExistingStreamError
   */
  def unsubscribe(subscriptions: String, principals: String)
      (implicit toEntityMarshallerNonExistingStreamError: ToEntityMarshaller[NonExistingStreamError], toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route

  def updateStream200(responseJsonSuccess: JsonSuccess)(implicit toEntityMarshallerJsonSuccess: ToEntityMarshaller[JsonSuccess]): Route =
    complete((200, responseJsonSuccess))
  def updateStream400(responseJsonError: JsonError)(implicit toEntityMarshallerJsonError: ToEntityMarshaller[JsonError]): Route =
    complete((400, responseJsonError))
  /**
   * Code: 200, Message: Success., DataType: JsonSuccess
   * Code: 400, Message: Bad request., DataType: JsonError
   */
  def updateStream(streamId: Int, description: Option[String], newName: Option[String], isPrivate: Option[Boolean], isAnnouncementOnly: Option[Boolean], streamPostPolicy: Int, historyPublicToSubscribers: Option[Boolean], messageRetentionDays: Option[String])
      (implicit toEntityMarshallerJsonSuccess: ToEntityMarshaller[JsonSuccess], toEntityMarshallerJsonError: ToEntityMarshaller[JsonError]): Route

  def updateSubscriptionSettings200(responseJsonSuccessBase: JsonSuccessBase)(implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route =
    complete((200, responseJsonSuccessBase))
  /**
   * Code: 200, Message: Success., DataType: JsonSuccessBase
   */
  def updateSubscriptionSettings(subscriptionData: String)
      (implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route

  def updateSubscriptions200(responseJsonSuccessBase: JsonSuccessBase)(implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route =
    complete((200, responseJsonSuccessBase))
  /**
   * Code: 200, Message: Success., DataType: JsonSuccessBase
   */
  def updateSubscriptions(delete: String, add: String)
      (implicit toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]): Route

}

trait StreamsApiMarshaller {


  implicit def toEntityMarshallerNonExistingStreamError: ToEntityMarshaller[NonExistingStreamError]

  implicit def toEntityMarshallerOneOfobjectobject: ToEntityMarshaller[OneOfobjectobject]

  implicit def toEntityMarshallerCodedError: ToEntityMarshaller[CodedError]

  implicit def toEntityMarshallerJsonSuccess: ToEntityMarshaller[JsonSuccess]

  implicit def toEntityMarshallerJsonSuccessBase: ToEntityMarshaller[JsonSuccessBase]

  implicit def toEntityMarshallerJsonError: ToEntityMarshaller[JsonError]

}

File: dart-jaguar/lib/api/streams_api.dart
import 'package:jaguar_retrofit/annotations/annotations.dart';
import 'package:jaguar_retrofit/jaguar_retrofit.dart';
import 'package:jaguar_serializer/jaguar_serializer.dart';
import 'package:jaguar_mimetype/jaguar_mimetype.dart';
import 'dart:async';

import 'package:openapi/model/json_success_base.dart';
import 'package:openapi/model/non_existing_stream_error.dart';
import 'package:openapi/model/json_success.dart';
import 'package:openapi/model/one_ofstringinteger.dart';
import 'package:openapi/model/one_ofobjectobject.dart';
import 'package:openapi/model/json_error.dart';
import 'package:openapi/model/coded_error.dart';

part 'streams_api.jretro.dart';

@GenApiClient()
class StreamsApi extends ApiClient with _$StreamsApiClient {
    final Route base;
    final Map<String, CodecRepo> converters;
    final Duration timeout;

    StreamsApi({this.base, this.converters, this.timeout = const Duration(minutes: 2)});

    /// Archive a stream
    ///
    /// [Archive the stream](/help/archive-a-stream) with the ID &#x60;stream_id&#x60;.  &#x60;DELETE {{ api_url }}/v1/streams/{stream_id}&#x60; 
    @DeleteReq(path: "/streams/:stream_id")
    Future<JsonSuccess> archiveStream(
            @PathParam("stream_id") int streamId
        ) {
        return super.archiveStream(
        streamId

        ).timeout(timeout);
    }

    /// Create BigBlueButton video call
    ///
    /// Create a video call URL for a BigBlueButton video call. Requires BigBlueButton to be configured on the Zulip server. 
    @GetReq(path: "/calls/bigbluebutton/create")
    Future<JsonSuccessBase> createBigBlueButtonVideoCall(
        ) {
        return super.createBigBlueButtonVideoCall(

        ).timeout(timeout);
    }

    /// Delete a topic
    ///
    /// Delete all messages in a topic.  &#x60;POST {{ api_url }}/v1/streams/{stream_id}/delete_topic&#x60;  Topics are a field on messages (not an independent data structure), so deleting all the messages in the topic deletes the topic from Zulip. 
    @PostReq(path: "/streams/:stream_id/delete_topic")
    Future<JsonSuccess> deleteTopic(
            @PathParam("stream_id") int streamId
        ,
            @QueryParam("topic_name") String topicName
        ) {
        return super.deleteTopic(
        streamId
        ,
        topicName

        ).timeout(timeout);
    }

    /// Get stream ID
    ///
    /// Get the unique ID of a given stream.  &#x60;GET {{ api_url }}/v1/get_stream_id&#x60; 
    @GetReq(path: "/get_stream_id")
    Future<JsonSuccessBase> getStreamId(
        
            @QueryParam("stream") String stream
        ) {
        return super.getStreamId(
        
        stream

        ).timeout(timeout);
    }

    /// Get topics in a stream
    ///
    /// Get all the topics in a specific stream  &#x60;GET {{ api_url }}/v1/users/me/{stream_id}/topics&#x60; 
    @GetReq(path: "/users/me/:stream_id/topics")
    Future<JsonSuccessBase> getStreamTopics(
            @PathParam("stream_id") int streamId
        ) {
        return super.getStreamTopics(
        streamId

        ).timeout(timeout);
    }

    /// Get all streams
    ///
    /// Get all streams that the user has access to.  &#x60;GET {{ api_url }}/v1/streams&#x60; 
    @GetReq(path: "/streams")
    Future<JsonSuccessBase> getStreams(
        
            @QueryParam("include_public") bool includePublic, 
        
            @QueryParam("include_web_public") bool includeWebPublic, 
        
            @QueryParam("include_subscribed") bool includeSubscribed, 
        
            @QueryParam("include_all_active") bool includeAllActive, 
        
            @QueryParam("include_default") bool includeDefault, 
        
            @QueryParam("include_owner_subscribed") bool includeOwnerSubscribed
        ) {
        return super.getStreams(
        
        includePublic, 
        
        includeWebPublic, 
        
        includeSubscribed, 
        
        includeAllActive, 
        
        includeDefault, 
        
        includeOwnerSubscribed

        ).timeout(timeout);
    }

    /// Get the subscribers of a stream
    ///
    /// Get all users subscribed to a stream.  &#x60;Get {{ api_url }}/v1/streams/{stream_id}/members&#x60; 
    @GetReq(path: "/streams/:stream_id/members")
    Future<JsonSuccessBase> getSubscribers(
            @PathParam("stream_id") int streamId
        ) {
        return super.getSubscribers(
        streamId

        ).timeout(timeout);
    }

    /// Get subscription status
    ///
    /// Check whether a user is subscribed to a stream.  &#x60;GET {{ api_url }}/v1/users/{user_id}/subscriptions/{stream_id}&#x60;  **Changes**: New in Zulip 3.0 (feature level 11). 
    @GetReq(path: "/users/:user_id/subscriptions/:stream_id")
    Future<JsonSuccessBase> getSubscriptionStatus(
            @PathParam("user_id") int userId, 
            @PathParam("stream_id") int streamId
        ) {
        return super.getSubscriptionStatus(
        userId, 
        streamId

        ).timeout(timeout);
    }

    /// Get subscribed streams
    ///
    /// Get all streams that the user is subscribed to.  &#x60;GET {{ api_url }}/v1/users/me/subscriptions&#x60; 
    @GetReq(path: "/users/me/subscriptions")
    Future<JsonSuccessBase> getSubscriptions(
        
            @QueryParam("include_subscribers") bool includeSubscribers
        ) {
        return super.getSubscriptions(
        
        includeSubscribers

        ).timeout(timeout);
    }

    /// Topic muting
    ///
    /// This endpoint mutes/unmutes a topic within a stream that the current user is subscribed to.  Muted topics are displayed faded in the Zulip UI, and are not included in the user&#39;s unread count totals.  &#x60;PATCH {{ api_url }}/v1/users/me/subscriptions/muted_topics&#x60; 
    @PatchReq(path: "/users/me/subscriptions/muted_topics")
    Future<JsonSuccess> muteTopic(
        
            @QueryParam("stream") String stream, 
        
            @QueryParam("stream_id") int streamId, 
        
            @QueryParam("topic") String topic, 
        
            @QueryParam("op") String op
        ) {
        return super.muteTopic(
        
        stream, 
        
        streamId, 
        
        topic, 
        
        op

        ).timeout(timeout);
    }

    /// Subscribe to a stream
    ///
    /// Subscribe one or more users to one or more streams.  &#x60;POST {{ api_url }}/v1/users/me/subscriptions&#x60;  If any of the specified streams do not exist, they are automatically created.  The initial [stream settings](/api/update-stream) will be determined by the optional parameters like &#x60;invite_only&#x60; detailed below. 
    @PostReq(path: "/users/me/subscriptions")
    Future<OneOfobjectobject> subscribe(
        
            @QueryParam("subscriptions") List<Object> subscriptions, 
        
            @QueryParam("principals") List<OneOfstringinteger> principals, 
        
            @QueryParam("authorization_errors_fatal") bool authorizationErrorsFatal, 
        
            @QueryParam("announce") bool announce, 
        
            @QueryParam("invite_only") bool inviteOnly, 
        
            @QueryParam("history_public_to_subscribers") bool historyPublicToSubscribers, 
        
            @QueryParam("stream_post_policy") int streamPostPolicy, 
        
            @QueryParam("message_retention_days") OneOfstringinteger messageRetentionDays
        ) {
        return super.subscribe(
        
        subscriptions, 
        
        principals, 
        
        authorizationErrorsFatal, 
        
        announce, 
        
        inviteOnly, 
        
        historyPublicToSubscribers, 
        
        streamPostPolicy, 
        
        messageRetentionDays

        ).timeout(timeout);
    }

    /// Unsubscribe from a stream
    ///
    /// Unsubscribe yourself or other users from one or more streams.  &#x60;DELETE {{ api_url }}/v1/users/me/subscriptions&#x60; 
    @DeleteReq(path: "/users/me/subscriptions")
    Future<JsonSuccessBase> unsubscribe(
        
            @QueryParam("subscriptions") List<String> subscriptions, 
        
            @QueryParam("principals") List<OneOfstringinteger> principals
        ) {
        return super.unsubscribe(
        
        subscriptions, 
        
        principals

        ).timeout(timeout);
    }

    /// Update a stream
    ///
    /// Configure the stream with the ID &#x60;stream_id&#x60;.  This endpoint supports an organization administrator editing any property of a stream, including:  * Stream [name](/help/rename-a-stream) and [description](/help/change-the-stream-description) * Stream [permissions](/help/stream-permissions), including [privacy](/help/change-the-privacy-of-a-stream) and [who can send](/help/stream-sending-policy).  &#x60;PATCH {{ api_url }}/v1/streams/{stream_id}&#x60; 
    @PatchReq(path: "/streams/:stream_id")
    Future<JsonSuccess> updateStream(
            @PathParam("stream_id") int streamId
        ,
            @QueryParam("description") String description, 
        
            @QueryParam("new_name") String newName, 
        
            @QueryParam("is_private") bool isPrivate, 
        
            @QueryParam("is_announcement_only") bool isAnnouncementOnly, 
        
            @QueryParam("stream_post_policy") int streamPostPolicy, 
        
            @QueryParam("history_public_to_subscribers") bool historyPublicToSubscribers, 
        
            @QueryParam("message_retention_days") OneOfstringinteger messageRetentionDays
        ) {
        return super.updateStream(
        streamId
        ,
        description, 
        
        newName, 
        
        isPrivate, 
        
        isAnnouncementOnly, 
        
        streamPostPolicy, 
        
        historyPublicToSubscribers, 
        
        messageRetentionDays

        ).timeout(timeout);
    }

    /// Update subscription settings
    ///
    /// This endpoint is used to update the user&#39;s personal settings for the streams they are subscribed to, including muting, color, pinning, and per-stream notification settings.  &#x60;POST {{ api_url }}/v1/users/me/subscriptions/properties&#x60; 
    @PostReq(path: "/users/me/subscriptions/properties")
    Future<JsonSuccessBase> updateSubscriptionSettings(
        
            @QueryParam("subscription_data") List<Object> subscriptionData
        ) {
        return super.updateSubscriptionSettings(
        
        subscriptionData

        ).timeout(timeout);
    }

    /// Update subscriptions
    ///
    /// Update which streams you are are subscribed to. 
    @PatchReq(path: "/users/me/subscriptions")
    Future<JsonSuccessBase> updateSubscriptions(
        
            @QueryParam("delete") List<String> delete, 
        
            @QueryParam("add") List<Object> add
        ) {
        return super.updateSubscriptions(
        
        delete, 
        
        add

        ).timeout(timeout);
    }


}
